File: app/src/main/java/org/caojun/salmagundi/utils/DataStorageUtils.java
package org.caojun.salmagundi.utils;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.text.TextUtils;

/**
 * 存储工具类
 * Created by CaoJun on 2016/11/3.
 */

public class DataStorageUtils {

    private static boolean isIllegal(Context context, String name, String key)
    {
        return context == null || TextUtils.isEmpty(name) || TextUtils.isEmpty(key);
    }

    public static Editor getEditor(Context context, String name, String key)
    {
        if(isIllegal(context, name, key))
        {
            return null;
        }
        SharedPreferences preferences = context.getSharedPreferences(name, Context.MODE_PRIVATE);
        return preferences.edit();
    }

    public static SharedPreferences getSharedPreferences(Context context, String name, String key)
    {
        if(isIllegal(context, name, key))
        {
            return null;
        }
        return context.getSharedPreferences(name, Context.MODE_PRIVATE);
    }

    public static boolean saveBoolean(Context context, String name, String key, boolean value)
    {
        Editor editor = getEditor(context, name, key);
        if(editor == null)
        {
            return false;
        }
        editor.putBoolean(key, value);
        editor.commit();
        return true;
    }

    public static Boolean loadBoolean(Context context, String name, String key, boolean defValue)
    {
        SharedPreferences preferences = getSharedPreferences(context, name, key);
        if(preferences == null)
        {
            return null;
        }
        return preferences.getBoolean(key, defValue);
    }

    public static boolean saveBooleanArray(Context context, String name, String key, boolean[] values)
    {
        Editor editor = getEditor(context, name, key);
        if(editor == null || values == null)
        {
            return false;
        }
        editor.putInt(key + ".length", values.length);
        for(int i = 0;i < values.length;i ++)
        {
            editor.putBoolean(key + "." + i, values[i]);
        }
        editor.commit();
        return true;
    }

    public static Boolean[] loadBooleanArray(Context context, String name, String key, boolean defValue)
    {
        SharedPreferences preferences = getSharedPreferences(context, name, key);
        if(preferences == null)
        {
            return null;
        }
        int length = preferences.getInt(key + ".length", 0);
        if(length < 1)
        {
            return null;
        }
        Boolean[] values = new Boolean[length];
        for(int i = 0;i < length;i ++)
        {
            values[i] = preferences.getBoolean(key + "." + i, defValue);
        }
        return values;
    }

    public static boolean saveFloat(Context context, String name, String key, float value)
    {
        Editor editor = getEditor(context, name, key);
        if(editor == null)
        {
            return false;
        }
        editor.putFloat(key, value);
        editor.commit();
        return true;
    }

    public static Float loadFloat(Context context, String name, String key, float defValue)
    {
        SharedPreferences preferences = getSharedPreferences(context, name, key);
        if(preferences == null)
        {
            return null;
        }
        return preferences.getFloat(key, defValue);
    }

    public static boolean saveFloatArray(Context context, String name, String key, float[] values)
    {
        Editor editor = getEditor(context, name, key);
        if(editor == null || values == null)
        {
            return false;
        }
        editor.putInt(key + ".length", values.length);
        for(int i = 0;i < values.length;i ++)
        {
            editor.putFloat(key + "." + i, values[i]);
        }
        editor.commit();
        return true;
    }

    public static Float[] loadFloatArray(Context context, String name, String key, float defValue)
    {
        SharedPreferences preferences = getSharedPreferences(context, name, key);
        if(preferences == null)
        {
            return null;
        }
        int length = preferences.getInt(key + ".length", 0);
        if(length < 1)
        {
            return null;
        }
        Float[] values = new Float[length];
        for(int i = 0;i < length;i ++)
        {
            values[i] = preferences.getFloat(key + "." + i, defValue);
        }
        return values;
    }

    public static boolean saveInt(Context context, String name, String key, int value)
    {
        Editor editor = getEditor(context, name, key);
        if(editor == null)
        {
            return false;
        }
        editor.putInt(key, value);
        editor.commit();
        return true;
    }

    public static Integer loadInt(Context context, String name, String key, int defValue)
    {
        SharedPreferences preferences = getSharedPreferences(context, name, key);
        if(preferences == null)
        {
            return null;
        }
        return preferences.getInt(key, defValue);
    }

    public static boolean saveIntArray(Context context, String name, String key, int[] values)
    {
        Editor editor = getEditor(context, name, key);
        if(editor == null || values == null)
        {
            return false;
        }
        editor.putInt(key + ".length", values.length);
        for(int i = 0;i < values.length;i ++)
        {
            editor.putInt(key + "." + i, values[i]);
        }
        editor.commit();
        return true;
    }

    public static boolean saveIntArray(Context context, String name, String key, Integer[] values)
    {
        Editor editor = getEditor(context, name, key);
        if(editor == null || values == null)
        {
            return false;
        }
        editor.putInt(key + ".length", values.length);
        for(int i = 0;i < values.length;i ++)
        {
            editor.putInt(key + "." + i, values[i]);
        }
        editor.commit();
        return true;
    }

    public static Integer[] loadIntArray(Context context, String name, String key, int defValue)
    {
        SharedPreferences preferences = getSharedPreferences(context, name, key);
        if(preferences == null)
        {
            return null;
        }
        int length = preferences.getInt(key + ".length", 0);
        if(length < 1)
        {
            return null;
        }
        Integer[] values = new Integer[length];
        for(int i = 0;i < length;i ++)
        {
            values[i] = preferences.getInt(key + "." + i, defValue);
        }
        return values;
    }

    public static boolean saveLong(Context context, String name, String key, long value)
    {
        Editor editor = getEditor(context, name, key);
        if(editor == null)
        {
            return false;
        }
        editor.putLong(key, value);
        editor.commit();
        return true;
    }

    public static Long loadLong(Context context, String name, String key, long defValue)
    {
        SharedPreferences preferences = getSharedPreferences(context, name, key);
        if(preferences == null)
        {
            return null;
        }
        return preferences.getLong(key, defValue);
    }

    public static boolean saveLongArray(Context context, String name, String key, long[] values)
    {
        Editor editor = getEditor(context, name, key);
        if(editor == null || values == null)
        {
            return false;
        }
        editor.putInt(key + ".length", values.length);
        for(int i = 0;i < values.length;i ++)
        {
            editor.putLong(key + "." + i, values[i]);
        }
        editor.commit();
        return true;
    }

    public static Long[] loadLongArray(Context context, String name, String key, long defValue)
    {
        SharedPreferences preferences = getSharedPreferences(context, name, key);
        if(preferences == null)
        {
            return null;
        }
        int length = preferences.getInt(key + ".length", 0);
        if(length < 1)
        {
            return null;
        }
        Long[] values = new Long[length];
        for(int i = 0;i < length;i ++)
        {
            values[i] = preferences.getLong(key + "." + i, defValue);
        }
        return values;
    }

    public static boolean saveString(Context context, String name, String key, String value)
    {
        Editor editor = getEditor(context, name, key);
        if(editor == null)
        {
            return false;
        }
        editor.putString(key, value);
        editor.commit();
        return true;
    }

    public static String loadString(Context context, String name, String key, String defValue)
    {
        SharedPreferences preferences = getSharedPreferences(context, name, key);
        if(preferences == null)
        {
            return null;
        }
        return preferences.getString(key, defValue);
    }

    public static boolean saveStringArray(Context context, String name, String key, String[] values)
    {
        Editor editor = getEditor(context, name, key);
        if(editor == null || values == null)
        {
            return false;
        }
        editor.putInt(key + ".length", values.length);
        for(int i = 0;i < values.length;i ++)
        {
            editor.putString(key + "." + i, values[i]);
        }
        editor.commit();
        return true;
    }

    public static String[] loadStringArray(Context context, String name, String key, String defValue)
    {
        SharedPreferences preferences = getSharedPreferences(context, name, key);
        if(preferences == null)
        {
            return null;
        }
        int length = preferences.getInt(key + ".length", 0);
        if(length < 1)
        {
            return null;
        }
        String[] values = new String[length];
        for(int i = 0;i < length;i ++)
        {
            values[i] = preferences.getString(key + "." + i, defValue);
        }
        return values;
    }

    public static boolean saveByte(Context context, String name, String key, byte value)
    {
        Editor editor = getEditor(context, name, key);
        if(editor == null)
        {
            return false;
        }
        editor.putInt(key, value);
        editor.commit();
        return true;
    }

    public static Byte loadByte(Context context, String name, String key, byte defValue)
    {
        SharedPreferences preferences = getSharedPreferences(context, name, key);
        if(preferences == null)
        {
            return null;
        }
        Integer value = preferences.getInt(key, defValue);
        return Byte.valueOf(value.byteValue());
    }

    public static boolean saveByteArray(Context context, String name, String key, byte[] values)
    {
        Editor editor = getEditor(context, name, key);
        if(editor == null || values == null)
        {
            return false;
        }
        editor.putInt(key + ".length", values.length);
        for(int i = 0;i < values.length;i ++)
        {
            editor.putInt(key + "." + i, values[i]);
        }
        editor.commit();
        return true;
    }

    public static byte[] loadByteArray(Context context, String name, String key, byte defValue)
    {
        SharedPreferences preferences = getSharedPreferences(context, name, key);
        if(preferences == null)
        {
            return null;
        }
        int length = preferences.getInt(key + ".length", 0);
        if(length < 1)
        {
            return null;
        }
        byte[] values = new byte[length];
        for(int i = 0;i < length;i ++)
        {
            Integer value = preferences.getInt(key + "." + i, defValue);
            values[i] = Byte.valueOf(value.byteValue());
        }
        return values;
    }
}

File: UtilsLibrary/src/main/java/org/caojun/utils/DataStorageUtils.kt
package org.caojun.utils

import android.content.Context
import android.content.SharedPreferences
import android.graphics.Bitmap
import android.preference.PreferenceManager
import android.text.TextUtils

/**
 * Created by CaoJun on 2017/9/27.
 */
object DataStorageUtils {

    private fun isIllegal(context: Context?, key: String): Boolean {
        return context == null || TextUtils.isEmpty(key)
    }

    fun getEditor(context: Context): SharedPreferences.Editor {
        return PreferenceManager.getDefaultSharedPreferences(context).edit()
    }

    fun getSharedPreferences(context: Context): SharedPreferences {
        return PreferenceManager.getDefaultSharedPreferences(context)
    }

    fun saveBoolean(context: Context, key: String, value: Boolean): Boolean {
        val editor = getEditor(context)
        editor.putBoolean(key, value)
        editor.commit()
        return true
    }

    fun loadBoolean(context: Context, key: String, defValue: Boolean): Boolean {
        val sp = getSharedPreferences(context)
        return sp.getBoolean(key, defValue)
    }

    fun saveBooleanArray(context: Context, key: String, values: BooleanArray): Boolean {
        val editor = getEditor(context)
//        editor.putInt(key + ".length", values.size)
        editor.putInt("$key.length", values.size)
        for (i in values.indices) {
//            editor.putBoolean(key + "." + i, values[i])
            editor.putBoolean("$key.$i", values[i])
        }
        editor.commit()
        return true
    }

    fun loadBooleanArray(context: Context, key: String, defValue: Boolean): Array<Boolean?>? {
        val preferences = getSharedPreferences(context)
//        val length = preferences.getInt(key + ".length", 0)
        val length = preferences.getInt("$key.length", 0)
        if (length < 1) {
            return null
        }
        val values = arrayOfNulls<Boolean>(length)
        for (i in 0 until length) {
//            values[i] = preferences.getBoolean(key + "." + i, defValue)
            values[i] = preferences.getBoolean("$key.$i", defValue)
        }
        return values
    }

    fun saveFloat(context: Context, key: String, value: Float): Boolean {
        val editor = getEditor(context)
        editor.putFloat(key, value)
        editor.commit()
        return true
    }

    fun loadFloat(context: Context, key: String, defValue: Float): Float {
        val preferences = getSharedPreferences(context)
        return preferences.getFloat(key, defValue)
    }

    fun saveFloatArray(context: Context, key: String, values: FloatArray): Boolean {
        val editor = getEditor(context)
//        editor.putInt(key + ".length", values.size)
        editor.putInt("$key.length", values.size)
        for (i in values.indices) {
//            editor.putFloat(key + "." + i, values[i])
            editor.putFloat("$key.$i", values[i])
        }
        editor.commit()
        return true
    }

    fun loadFloatArray(context: Context, key: String, defValue: Float): Array<Float?>? {
        val preferences = getSharedPreferences(context)
//        val length = preferences.getInt(key + ".length", 0)
        val length = preferences.getInt("$key.length", 0)
        if (length < 1) {
            return null
        }
        val values = arrayOfNulls<Float>(length)
        for (i in 0 until length) {
//            values[i] = preferences.getFloat(key + "." + i, defValue)
            values[i] = preferences.getFloat("$key.$i", defValue)
        }
        return values
    }

    fun saveInt(context: Context, key: String, value: Int): Boolean {
        val editor = getEditor(context)
        editor.putInt(key, value)
        editor.commit()
        return true
    }

    fun loadInt(context: Context, key: String, defValue: Int): Int {
        val preferences = getSharedPreferences(context)
        return preferences.getInt(key, defValue)
    }

    fun saveIntArray(context: Context, key: String, values: IntArray): Boolean {
        val editor = getEditor(context)
//        editor.putInt(key + ".length", values.size)
        editor.putInt("$key.length", values.size)
        for (i in values.indices) {
//            editor.putInt(key + "." + i, values[i])
            editor.putInt("$key.$i", values[i])
        }
        editor.commit()
        return true
    }

    fun saveIntArray(context: Context, key: String, values: Array<Int>): Boolean {
        val editor = getEditor(context)
//        editor.putInt(key + ".length", values.size)
        editor.putInt("$key.length", values.size)
        for (i in values.indices) {
//            editor.putInt(key + "." + i, values[i])
            editor.putInt("$key.$i", values[i])
        }
        editor.commit()
        return true
    }

    fun loadIntArray(context: Context, key: String, defValue: Int): Array<Int?>? {
        val preferences = getSharedPreferences(context)
//        val length = preferences.getInt(key + ".length", 0)
        val length = preferences.getInt("$key.length", 0)
        if (length < 1) {
            return null
        }
        val values = arrayOfNulls<Int>(length)
        for (i in 0 until length) {
//            values[i] = preferences.getInt(key + "." + i, defValue)
            values[i] = preferences.getInt("$key.$i", defValue)
        }
        return values
    }

    fun saveLong(context: Context, key: String, value: Long): Boolean {
        val editor = getEditor(context)
        editor.putLong(key, value)
        editor.commit()
        return true
    }

    fun loadLong(context: Context, key: String, defValue: Long): Long {
        val preferences = getSharedPreferences(context)
        return preferences.getLong(key, defValue)
    }

    fun saveLongArray(context: Context, key: String, values: LongArray): Boolean {
        val editor = getEditor(context)
//        editor.putInt(key + ".length", values.size)
        editor.putInt("$key.length", values.size)
        for (i in values.indices) {
//            editor.putLong(key + "." + i, values[i])
            editor.putLong("$key.$i", values[i])
        }
        editor.commit()
        return true
    }

    fun loadLongArray(context: Context, key: String, defValue: Long): Array<Long?>? {
        val preferences = getSharedPreferences(context)
//        val length = preferences.getInt(key + ".length", 0)
        val length = preferences.getInt("$key.length", 0)
        if (length < 1) {
            return null
        }
        val values = arrayOfNulls<Long>(length)
        for (i in 0 until length) {
//            values[i] = preferences.getLong(key + "." + i, defValue)
            values[i] = preferences.getLong("$key.$i", defValue)
        }
        return values
    }

    fun saveString(context: Context, key: String, value: String): Boolean {
        val editor = getEditor(context)
        editor.putString(key, value)
        editor.commit()
        return true
    }

    fun loadString(context: Context, key: String, defValue: String): String {
        val preferences = getSharedPreferences(context)
        return preferences.getString(key, defValue)
    }

    fun saveStringArray(context: Context, key: String, values: Array<String>): Boolean {
        val editor = getEditor(context)
//        editor.putInt(key + ".length", values.size)
        editor.putInt("$key.length", values.size)
        for (i in values.indices) {
//            editor.putString(key + "." + i, values[i])
            editor.putString("$key.$i", values[i])
        }
        editor.commit()
        return true
    }

    fun loadStringArray(context: Context, key: String, defValue: String): Array<String?>? {
        val preferences = getSharedPreferences(context)
//        val length = preferences.getInt(key + ".length", 0)
        val length = preferences.getInt("$key.length", 0)
        if (length < 1) {
            return null
        }
        val values = arrayOfNulls<String>(length)
        for (i in 0 until length) {
//            values[i] = preferences.getString(key + "." + i, defValue)
            values[i] = preferences.getString("$key.$i", defValue)
        }
        return values
    }

    fun saveByte(context: Context, key: String, value: Byte): Boolean {
        val editor = getEditor(context)
        editor.putInt(key, value.toInt())
        editor.commit()
        return true
    }

    fun loadByte(context: Context, key: String, defValue: Byte): Byte {
        val preferences = getSharedPreferences(context)
        val value = preferences.getInt(key, defValue.toInt())
        return java.lang.Byte.valueOf(value.toByte())
    }

    fun saveByteArray(context: Context, key: String, values: ByteArray?): Boolean {
        val editor = getEditor(context)
//        editor.putInt(key + ".length", values?.size?:0)
        editor.putInt("$key.length", values?.size?:0)
        if (values != null) {
            for (i in values.indices) {
//                editor.putInt(key + "." + i, values[i].toInt())
                editor.putInt("$key.$i", values[i].toInt())
            }
        }
        editor.commit()
        return true
    }

    fun loadByteArray(context: Context, key: String, defValue: Byte): ByteArray? {
        val preferences = getSharedPreferences(context)
//        val length = preferences.getInt(key + ".length", 0)
        val length = preferences.getInt("$key.length", 0)
        if (length < 1) {
            return null
        }
        val values = ByteArray(length)
        for (i in 0 until length) {
//            val value = preferences.getInt(key + "." + i, defValue.toInt())
            val value = preferences.getInt("$key.$i", defValue.toInt())
            values[i] = java.lang.Byte.valueOf(value.toByte())!!
        }
        return values
    }

    fun saveBitmap(context: Context, key: String, bitmap: Bitmap?): Boolean {
        val data = ImageUtils.toByteArray(bitmap)
        return saveByteArray(context, key, data)
    }

    fun loadBitmap(context: Context, key: String, defValue: Bitmap?): Bitmap? {
        val data = loadByteArray(context, key, 0)
        return ImageUtils.toBitmap(data) ?: defValue
    }
}